### PR TITLE
[dif] Implement DIFs for alert_handler

### PR DESCRIPTION
**Pull request by mcy**
Unit tests coming soon, ~~probably as part of this PR.~~

### CODE REVIEW

**Comment by silvestrst** [comment]
Probably was missed in the header review:
`dif_alert_handler.h:L{261, 280}`, `See dif_alert_manager_class_config_t` => `See dif_alert_handler_class_config_t`

**Comment by silvestrst** [comment]
In `dif_alert_handler.h:L303`:
```
* For each of the pointer/length field pairs below, if the length is zero, the 
* pointer may be `NULL`.
```
It seems like a slightly "special" case from when `length > 0`. Could you please document an use-case for `length == 0`? For example, when the `length == 0` it skips the big configuration `for` loop entirely.

**Review by silvestrst — COMMENTED** [review]
@mcy thank you!

This is the first pass review, I intend to come back and re-review this PR either later this week or next week.

So far I only have made my way through:
`dif_alert_handler_init`
`classify_alerts`
`classify_local_alerts`

and partially

`configure_class`

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:23]: That value can vary with the hardware instantiation; the DIF assumes that that value is 4, which we've already static asserted. I feel like a literal is more clear than the generated constant, since it can't change.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:34]: No, "any" here is as in "any part of the provided configuration".

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:37]: Is distinguishing those failures useful? Note that returning kBadArg is incorrect, since this function may be called after hardware has been touched.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:76]: I think we currently assume no more than 16 alerts, right? @msfschaffner do you think this is a reasonable assumption?

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:71]: I felt it was better to keep the definition closer to where it is used. This is a trivial case of loop-invariant code motion, so it's not going to get recomputed every time. I can move if you feel very strongly about it though.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:99]: See my other comment.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:230]: I intentionally kept it as once stanza, since this `if` statement validates that the for loop can be executed.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:42]: That feels a bit weird. I don't think it's worth introducing a new name just to save on characters.

**Comment by mcy** [comment]
> It seems like a slightly "special" case from when length > 0. Could you please document an use-case for length == 0? For example, when the length == 0 it skips the big configuration for loop entirely.

It feels perfectly valid to pass an empty list here? I felt it would be a good idea to be consistent, since very struct that carries such a pointer/len pair has values other than the pair that will still be used for things. (Requiring non-NULL for an empty list is a bit silly in my opinion).

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:286]: The resulting struct would be pretty big, and I'm not sure that LLVM can optimize those loads away. @lenary @gkelly thoughts?

File: sw/device/lib/dif/dif_alert_handler.c
Comment (silvestrst) [reply sw/device/lib/dif/dif_alert_handler.c:42]: Well, because this function feeds in into the public API, you won't be able to return a correct error code to the external caller. I suppose if we don't care about how precise the error codes returned to the user are, we could have just used `bool` as the return type of all the public API.

I personally think that if it fails in several ways, then it should return different error codes, but lets see what the other reviewers think.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (silvestrst) [reply sw/device/lib/dif/dif_alert_handler.c:78]: I don't have strong feelings, was just looking from the perspective of making the loop smaller. I think that simplifies it slightly, but arguably not much - it's up to you :)

File: sw/device/lib/dif/dif_alert_handler.c
Comment (lenary) [reply sw/device/lib/dif/dif_alert_handler.c:28]: This isn't the approach the PLIC takes, which we asked you to match. For instance, the PLIC DIF has a `static_assert` about the value of `RV_PLIC_PARAM_NUMSRC` - and then uses that define elsewhere. When there are top-changes, usually the only place that needs to change is the static assert, as we haven't tripped over any new boundaries - for instance, look at the DIF (and test) changes in #3501 - which are very minimal. If you use literals, it's not minimal, and it's easy to miss updating one, which is why we have the defines at all.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:42]: I think the important question to ask is "how can the caller respond"? I don't think it's useful to report a complicated error to the caller, because it's still going to be hard to figure out where the bad configuration is just from an enum.

In general, I think we've been running on the assumption that all the DIFs can only be usefully debugged with gdb et al.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:28]: I'm not sure that follows. It's not just the literals that need to be updated; there is logic that materially assumes that it's 4; for example, which assumes that only one register is defined to hold certain values, which would change from under the DIF if we just updated the `_Static_assert`.

I can change it, but I don't think it's a good idea.

Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:28]: Ah, ok. I think it makes more sense, in that case.

**Comment by silvestrst** [comment]
> > It seems like a slightly "special" case from when length > 0. Could you please document an use-case for length == 0? For example, when the length == 0 it skips the big configuration for loop entirely.
> 
> It feels perfectly valid to pass an empty list here? I felt it would be a good idea to be consistent, since very struct that carries such a pointer/len pair has values other than the pair that will still be used for things. (Requiring non-NULL for an empty list is a bit silly in my opinion).

It seems that with `class->alerts_len == 0` what you do is read the current value in the `enable` and `alerts` register:
```
  uint32_t enable_reg = mmio_region_read32(handler->params.base_addr,
                                           ALERT_HANDLER_ALERT_EN_REG_OFFSET);
  uint32_t alerts_reg = mmio_region_read32(
      handler->params.base_addr, ALERT_HANDLER_ALERT_CLASS_REG_OFFSET);
```
and then because `class->alerts_len == 0`, you skip the `for` loop and just write back the same values that you just read into the `enable` and `alert` registers, in:
```
  mmio_region_write32(handler->params.base_addr,
                      ALERT_HANDLER_ALERT_EN_REG_OFFSET, enable_reg);
  mmio_region_write32(handler->params.base_addr,
                      ALERT_HANDLER_ALERT_CLASS_REG_OFFSET, alerts_reg);
```

This seems unnecessary, and instead you could early return in such case, but please do correct me if I am wrong.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (silvestrst) [inline sw/device/lib/dif/dif_alert_handler.c:179]: It almost feels like this function could be split in multiple static functions (matter of preference I suppose).

File: sw/device/lib/dif/dif_alert_handler.c
Comment (silvestrst) [inline sw/device/lib/dif/dif_alert_handler.c:238]: Could you please add description for this loop?

File: sw/device/lib/dif/dif_alert_handler.c
Comment (silvestrst) [inline sw/device/lib/dif/dif_alert_handler.c:333]: Magic number `4`, please document or create a named define/constant.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (silvestrst) [inline sw/device/lib/dif/dif_alert_handler.c:364]: Magic number `4`, please document or create a named define/constant.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (silvestrst) [inline sw/device/lib/dif/dif_alert_handler.c:333]: I would have moved it out to the top of the file (IMHO) would make the loop easier to read (matter of preference I suppose).

File: sw/device/lib/dif/dif_alert_handler.c
Comment (silvestrst) [inline sw/device/lib/dif/dif_alert_handler.c:369]: Please correct me if I am wrong, but I thought the consensus was we don't check the validity of an input if it is an input is an enum defined in the DIF header (I think you might have suggested it yourself)?

**Review by silvestrst — COMMENTED** [review]
@mcy I had a second review pass on this PR, in which I have covered `configure_class` and `configure_phase_durations` functions.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:200]: @msfschaffner

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:333]: I don't believe the second one has a convenient constant anywhere, so I'll just document it.

Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:333]: It's only used in one place, so I think the locality is valuable here. Putting helpers far from where they are used is an antipattern.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:369]: I'm not sure if that's true, but this is checking that it's within a subset of the enum (since this enum contains other potential class states that aren't escalation phases).

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:179]: Maybe, but I feel like that would break up the control flow even more. I'll add comments to deliniate the different parts of the function.

It just feels big because it has a lot of silly switching going on.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (msfschaffner) [reply sw/device/lib/dif/dif_alert_handler.c:200]: This is assumption is valid. The register layout is identical for all four classes:

https://github.com/lowRISC/opentitan/blob/55405b41a1384e989f7f4c162d90db3bf9992040/hw/ip/alert_handler/data/alert_handler.hjson.tpl#L265-L270

File: sw/device/lib/dif/dif_alert_handler.h
Comment (msfschaffner) [inline sw/device/lib/dif/dif_alert_handler.h:399]: typo around `handle.will`?

File: sw/device/lib/dif/dif_alert_handler.c
Comment (msfschaffner) [reply sw/device/lib/dif/dif_alert_handler.c:83]: No this number is bound to grow as we are adding more and more alerts to our security sensitive blocks. Eventually, the number of alerts is likely going to be in a similar order of magnitude as the IRQs.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (msfschaffner) [inline sw/device/lib/dif/dif_alert_handler.c:22]: We have 14 alerts in the system at this time, but this number is bound to grow since we are adding new IP and alerts to existing blocks. I would expect that this is going to be in a similar order of magnitude as the #IRQs.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (msfschaffner) [inline sw/device/lib/dif/dif_alert_handler.c:343]: this number I think is exposed through `ALERT_HANDLER_PARAM_N_PHASES`.

**Review by alphan — APPROVED** [review]
Thanks @mcy, LGTM mod some comments. Approving just in case if you decide to create a different PR for tests.

**Review by silvestrst — APPROVED** [review]
Thanks @mcy !

Looks good to me `-` the outstanding comments. I think simple comments that you agree with is worth resolving before merge, but anything that requires further discussion or complicated changes should go in a follow up. I am leaving it up to you to decide how to treat these.

**Comment by silvestrst** [comment]
One thing I have forgot to mention in the review - I assume there could be multiple causes at any time  in `dif_alert_handler_alert_is_cause` and local alert function. If this assumption is correct, could you explicitly mention it in a note of function description in the header file?

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:343]: Thanks!

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:218]: I think it's a bit wasteful, given that it only occurs in three places, but whatever.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:549]: Not sure why this is commented on? This snippet only occurs once.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:244]: Yup, that's right.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:63]: Not as I understand it. "There are four classes" is a pretty solid assumptiuon.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:438]: I think that's pushing on overkill DRY IMO.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:651]: No, because the `alert_*` functions are patterend after the `irq_*` functions. Hence, from `irq_acknowledge` we get `alert_acknowledge`.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (msfschaffner) [reply sw/device/lib/dif/dif_alert_handler.c:65]: That is correct, we do not have plans to change the number of classes.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (msfschaffner) [reply sw/device/lib/dif/dif_alert_handler.c:24]: The only thing that changes when more alerts are added are the number of fields in the
```
ALERT_HANDLER.ALERT_EN
ALERT_HANDLER.ALERT_CLASS
ALERT_HANDLER.ALERT_CAUSE
```
CSRs. 

So we have to make sure that we can access the fields in those CSRs in a parametric way, even when they start to spill over into multiple CSRs (this happens at 32 alerts for `ALERT_EN` and `ALERT_CAUSE`, and 16 alerts for `ALERT_CLASS`).

File: sw/device/lib/dif/dif_alert_handler.c
Comment (gkelly) [reply sw/device/lib/dif/dif_alert_handler.c:201]: You resolved the comment, but didn't respond or update the change. What was the outcome here?

**Review by gkelly — APPROVED** [review]
Can you respond to Silvesters' comment that you resolved before landing please, for closure on the question.

File: sw/device/lib/dif/dif_alert_handler.c
Comment (mcy) [reply sw/device/lib/dif/dif_alert_handler.c:201]: I think this was resolved in another comment chain, where I argued that returning the different kinds of errors wasn't super meaningful; seems you and Alphan +1'd that view.

**Pull request merged**
